ompi/mca/osc/ucx/osc_ucx_active_target.c

```diff
@@ -0,0 +1,264 @@
+            /* not sure if this is required */
+            ucp_worker_progress(mca_osc_ucx_component.ucp_worker);
+        }
+    } else {
+        module->start_group = NULL;
+    }
+
+    return OMPI_SUCCESS;
+}
+
+int ompi_osc_ucx_complete(struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    ucs_status_t status;
+    int ret = OMPI_SUCCESS;
+
+    if (module->epoch_type.access != START_COMPLETE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    module->epoch_type.access = NONE_EPOCH;
+
+    status = ucp_worker_flush(mca_osc_ucx_component.ucp_worker);
+    if (status != UCS_OK) {
+        return OMPI_ERROR;
+    }
+    module->global_ops_num = 0;
+    memset(module->per_target_ops_nums, 0,
+           sizeof(int) * ompi_comm_size(module->comm));
+
+    if (module->start_group != NULL) {
+        int i, size;
+        ompi_group_t *win_group = NULL;
+        int *ranks_in_grp = NULL, *ranks_in_win_grp = NULL;
+
+        size = ompi_group_size(module->start_group);
+        ranks_in_grp = malloc(sizeof(int) * size);
+        ranks_in_win_grp = malloc(sizeof(int) * ompi_comm_size(module->comm));
+
+        for (i = 0; i < size; i++) {
+            ranks_in_grp[i] = i;
+        }
+
+        ret = ompi_comm_group(module->comm, &win_group);
+        if (ret != OMPI_SUCCESS) {
+            return OMPI_ERROR;
+        }
+
+        ret = ompi_group_translate_ranks(module->start_group, size, ranks_in_grp,
+                                         win_group, ranks_in_win_grp);
+        if (ret != OMPI_SUCCESS) {
+            return OMPI_ERROR;
+        }
+
+        for (i = 0; i < size; i++) {
+            uint64_t remote_addr = (module->state_info_array)[ranks_in_win_grp[i]].addr + 2 * sizeof(uint64_t); /* write to state.complete_count on remote side */
+            ucp_rkey_h rkey = (module->state_info_array)[ranks_in_win_grp[i]].rkey;
+            ucp_ep_h ep = OSC_UCX_GET_EP(module->comm, ranks_in_win_grp[i]);
+            uint64_t result_value;
+            void *req = ucp_atomic_fetch_nb(ep, UCP_ATOMIC_FETCH_OP_FADD, 1, &result_value,
+                                            sizeof(uint64_t), remote_addr, rkey,
+                                            amo_null_completion);
+            if (req != NULL) {
+                ucp_request_release(req);
+            }
+        }
+
+        free(ranks_in_grp);
+        free(ranks_in_win_grp);
+        ompi_group_free(&win_group);
+
+        OBJ_RELEASE(module->start_group);
+        module->start_group = NULL;
+
+        module->state.post_count = 0;
+    }
+
+    return ret;
+}
+
+int ompi_osc_ucx_post(struct ompi_group_t *group, int assert, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    int ret = OMPI_SUCCESS;
+
+    if (module->epoch_type.exposure != NONE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    module->epoch_type.exposure = POST_WAIT_EPOCH;
```

Comment:
(now L232) fixed. I move the update to the end of the function, which should be the success case.